module system: better option pretty-printing (add quotes when necessary)

danbst opened the pull request:
Fixes https://github.com/NixOS/nixpkgs/issues/18764

Because of lack or regexps in Nix, the implementation is larger than expected. I'm open to suggestions on how to simplify it

danbst committed:
module system: better option pretty-printing (add quotes when necessary)

Fixes https://github.com/NixOS/nixpkgs/issues/18764

dezgeg commented:
There is actually an undocumented (for unknown reasons) `builtins.match` that's already used in nixpkgs.

danbst commented:
@dezgeg much thanks! created https://github.com/NixOS/nix/issues/1145

7c6f434c commented:
I think Travis Failure actually looked somewhat relevant…

lheckemann commented:
This should probably also include the `.` character in the regex? Would be nice if we could get this merged after fixing the manual rendering.

danbst committed:
options.nix showOption: include some hardcodes

Those are used in manual in a special way.
Also, fix manual generation.

danbst commented:
@lheckemann thanks for the ping! I've fixed the manual build and rebased on top of master

Out of added goodies, found there exist a sole option
```
services.xserver.windowManager."2bwm".enable
```
which is now correctly rendered (compare https://nixos.org/nixos/manual/options.html#opt-services.xserver.windowManager.2bwm.enable)

danbst commented:
@vcunat 
re 2bwm - you may be interested in this PR too

lheckemann commented on lib/options.nix:
```diff
@@ -127,7 +127,17 @@ rec {
+    # https://github.com/NixOS/nix/blob/a5e761dddb6b090b233aebe29dc30ebfbc058dab/src/libexpr/lexer.l#L87
+    quotesWhenNeeded = x:
+      if x == "<name>" || x == "<name?>" || x == "*"  # special cases needed for documentation
+          || builtins.match ''[a-zA-Z\_][a-zA-Z0-9\_\'\-]*'' x != null
```
Maybe add "." here? While there are no stock options that this would affect, option attr paths do sometimes have .s in them, e.g. `services.nginx.virtualHosts."www.example.com"`.

lheckemann replied:
oh, of course.

danbst commented:
closing as essentially same was already done
https://github.com/NixOS/nixpkgs/pull/36344
